clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java

```diff
@@ -524,4 +524,32 @@ public void failed(Throwable throwable) {
 			}
 		}));	
 	}
+
+    /**helper method to include more useful information in the ErrorResponse for a ProcessingException*/
+    private static <T> Optional<T> readResponseToCallback(Response response, ResponseCallback<T> callback, GenericType<T> typeToken) {
+        response.bufferEntity();
+        try {
+            T entity = response.readEntity(typeToken);
+            callback.onSuccess(entity);
+            return Optional.of(entity);
+        } catch (ProcessingException pe) {
+            callback.onError(new ErrorResponse(pe.getMessage() + "; cause: " + pe.getCause() + "; body: " + response.readEntity(String.class)));
+            return Optional.empty();
+        }
+    }
+
+    /**helper method to include more useful information in the ErrorResponse for a ProcessingException*/
+    private static <T> void readResponseToCallback(Response response, ResponseCallback<T> callback, Class<T> clazz) {
+        readResponseToCallback(response, callback, clazz, UnaryOperator.identity());
+    }
+
+    /**helper method to include more useful information in the ErrorResponse for a ProcessingException*/
+    private static <T> void readResponseToCallback(Response response, ResponseCallback<T> callback, Class<T> clazz, UnaryOperator<T> finisher) {
+	    response.bufferEntity();
+	    try {
+	        callback.onSuccess(finisher.apply(response.readEntity(clazz)));
```

Comment:
What is this goal of applying the UnaryOperator's identity function to the response entity?

Reply:
The markExperimentPoint function looks to be using a different `finisher`. Is passing the identity function just for getting around that unique case?

Reply:
Yes, exactly.  Otherwise I'd have to repeat the rest of the implementation of `readResponseToCallback()`.

Indeed, I probably should've taken it one step further so that all _three_ overloads of that method could flow through one common code path.  That would've required even more functional programming-style abstraction, though, and would've made the code harder to follow.

A bit of a judgment call.  :)

Reply:
ok I just wanted to make sure I was following that correctly. 👍